### PR TITLE
KCWI  PR4

### PR DESCRIPTION
**Pull request by rcooke-ast**
Updates to the KCWI branch covering the recent broader updates to the code. This PR includes the wavelength calibration files for all KCWI modes I have to hand. The calibrations are completely and correctly made with this PR. There are also a few starting elements of the science reduction in `reduce.py`, but this will be the main issue tackled in the next PR. So, there's no need to comment on that side of things at this stage.

This PR is already merged with #918, so there's no need to review it until after 918 is merged.

### CODE REVIEW

File: pypeit/alignframe.py [hunk 1]
```diff
@@ -1,152 +1,20 @@
...
-#        return self.pypeitImage
+import astropy.io.fits as fits
+
+from pypeit import masterframe, io
+from pypeit import ginga, msgs
+from pypeit.core import extract, arc, load
+from pypeit.par.pypeitpar import AlignPar
 
 
 class Alignment():
```
Comment (profxj) [inline pypeit/alignframe.py:20]: on the next KCWI PR, we should turn this into a `DataContainer`.

I can help.

File: pypeit/alignframe.py [hunk 2]
```diff
@@ -302,89 +159,120 @@ def generate_dict(self, align_prof):
...
                 if align_prof[sls][bar].SLITID != sl:
                     msgs.error("Alignment profiling failed to generate dictionary")
-                    alignprof[:, bar, sl] = align_prof[sls][bar].TRACE_SPAT
+                    self._alignprof[:, bar, sl] = align_prof[sls][bar].TRACE_SPAT
         # Return the profile information as a single dictionary
-        return dict(alignments=alignprof)
+        return
 
-    def save(self, outfile=None, overwrite=True):
+    def save(self, outfile, overwrite=True, checksum=True, float_dtype='float32',
```
Comment (profxj) [inline pypeit/alignframe.py:176]: will become obsolete once we head to a DataContainer.

But ok for now.

File: pypeit/calibrations.py [hunk 3]
```diff
@@ -853,7 +842,7 @@ def default_steps():
 
         """
         # Order matters!
-        return ['bias', 'dark', 'bpm', 'slits', 'arc', 'tiltimg', 'wv_calib', 'tilts', 'flats']
+        return ['bias', 'dark', 'dark', 'bpm', 'slits', 'arc', 'tiltimg', 'wv_calib', 'tilts', 'flats']
```
Comment (profxj) [inline pypeit/calibrations.py:845]: there are now 2 entries of `dark`.  remove one.

File: pypeit/core/flat.py [hunk 4]
```diff
@@ -448,12 +448,483 @@ def tweak_slit_edges(left, right, spat_coo, norm_flat, thresh=0.93, maxfrac=0.1,
...
                                                spat_coo[i], right_thresh)
         msgs.info('Tweaking right slit boundary by {0:.1f}%'.format(100*right_shift) +
-                  ' % ({0:.2f} pixels)'.format(right_shift*slitwidth))        
+                  ' % ({0:.2f} pixels)'.format(right_shift*slitwidth))
         new_right -= right_shift * slitwidth
 
     return left_thresh, left_shift, new_left, right_thresh, right_shift, new_right
 
 
+# Deprecated.  Replaced by tweak_slit_edges above.
```
Comment (profxj) [inline pypeit/core/flat.py:457]: might as well delete

Comment (profxj) [reply pypeit/core/flat.py:457]: or move to deprecated/

File: pypeit/core/skysub.py [hunk 5]
```diff
@@ -1116,3 +1108,90 @@ def ech_local_skysub_extract(sciimg, sciivar, fullmask, tilts, waveimg, global_s
...
+def read_userregions(text, resolution=1000):
+    """
+
+    Args:
+        text: str
+            The sky region definition. The text should
+            be a comma separated list of percentages to apply to all slits
+            Example: The following string   :10,35:65,80:
+            would select the first 10%, the inner 30%, and the final 20% of all slits
+        resolution: int
```
Comment (profxj) [inline pypeit/core/skysub.py:1120]: optional

File: pypeit/core/skysub.py [hunk 6]
```diff
@@ -1116,3 +1108,90 @@ def ech_local_skysub_extract(sciimg, sciivar, fullmask, tilts, waveimg, global_s
     return skymodel, objmodel, ivarmodel, outmask, sobjs
 
 
+def read_userregions(text, resolution=1000):
+    """
+
+    Args:
```
Comment (profxj) [inline pypeit/core/skysub.py:1114]: This looks like a mix of `numpy` and `Google` docs.

File: pypeit/core/skysub.py [hunk 7]
```diff
@@ -1116,3 +1108,90 @@ def ech_local_skysub_extract(sciimg, sciivar, fullmask, tilts, waveimg, global_s
...
+                                round((resolution - 1) * float(tts[1]) / 100.0))
+                            ])
+    except:
+        status = 1
+    # Return
+    return status, regions
+
+
+def generate_mask(pypeline, regions, slits, slits_left, slits_right, resolution=1000):
+    """Generate the mask of sky regions
```
Comment (profxj) [inline pypeit/core/skysub.py:1154]: needs more docs

File: pypeit/core/skysub.py [hunk 8]
```diff
@@ -1116,3 +1108,90 @@ def ech_local_skysub_extract(sciimg, sciivar, fullmask, tilts, waveimg, global_s
...
+        if xmin < 0:
+            xmin = 0
+        if xmax > resolution:
+            xmax = resolution
+        # Apply to all slits
+        for sl in range(slits.nslits):
+            skyreg[sl][xmin:xmax] = 1
+
+    nreg = 0
+    left_edg, righ_edg = np.zeros((slits.nspec, 0)), np.zeros((slits.nspec, 0))
```
Comment (profxj) [inline pypeit/core/skysub.py:1176]: Please add a few more inline comments.  Hard to follow what is going on here (and why)

File: pypeit/core/wavecal/fitting.py [hunk 9]
```diff
@@ -102,7 +102,7 @@ def fit_slit(spec, patt_dict, tcent, line_lists, vel_tol = 1.0, outroot=None, sl
 # very least it should be optional
 def iterative_fitting(spec, tcent, ifit, IDs, llist, disp,
                       match_toler = 2.0, func = 'legendre', n_first=2, sigrej_first=2.0,
-                      n_final=4, sigrej_final=3.0,
+                      n_final=4, sigrej_final=3.0, input_only=False,
```
Comment (profxj) [inline pypeit/core/wavecal/fitting.py:105]: doc

File: pypeit/core/wavecal/waveio.py [hunk 10]
```diff
@@ -232,6 +232,12 @@ def load_line_list(line_file, add_path=False, use_ion=False, NIST=False):
     return line_list
 
 
+def get_available_lines(NIST=False):
```
Comment (profxj) [inline pypeit/core/wavecal/waveio.py:235]: doc string

File: pypeit/images/sciencecube.py [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+        pixel = Polygon([(0, 0), (0, 1), (1, 1), (1, 0)])
+        area = voxel.intersection(pixel).area
+
+    def build_crmask(self, subtract_img=None):
+        """
+        Call to ImageMask.build_crmask which will
+        generate the cosmic ray mask
+
+        Args:
+            subtract_img (np.ndarray, optional):
```
Comment (profxj) [inline pypeit/images/sciencecube.py:83]:
```
`numpy.ndarray`_
```
I am guilty of having put in a bunch of these erroneously..

File: pypeit/images/sciencecube.py [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+from pypeit.par import pypeitpar
+from pypeit import utils
+
+from pypeit.images import pypeitimage
+from pypeit.images import combineimage
+
+from IPython import embed
+
+
+class ScienceImage(pypeitimage.PypeItImage):
```
Comment (profxj) [inline pypeit/images/sciencecube.py:23]: Name `ScienceCube`

File: pypeit/images/sciencecube.py [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+            if getattr(self, attr) is not None:
+                rdict[attr] = True
+            else:
+                rdict[attr] = False
+        repr += ' images={}'.format(rdict)
+        repr = repr + '>'
+        return repr
+
+
+def build_from_file_list(spectrograph, det, par, bpm,
```
Comment (profxj) [inline pypeit/images/sciencecube.py:198]: I think this doesn't belong..

Maybe this whole module should be removed?

Comment (profxj) [reply pypeit/images/sciencecube.py:198]: If so, also the lines in `buildimages.py`

File: pypeit/reduce.py [hunk 14]
```diff
@@ -108,7 +115,7 @@ def get_instance(cls, sciImg, spectrograph, par, caliBrate,
 
     def __init__(self, sciImg, spectrograph, par, caliBrate,
                  objtype, ir_redux=False, det=1, std_redux=False, show=False,
-                 binning=None, setup=None):
+                 binning=None, setup=None, std_outfile=None):
```
Comment (profxj) [inline pypeit/reduce.py:118]: Add `std_outfile` to the doc string.

File: pypeit/reduce.py [hunk 15]
```diff
@@ -1081,3 +1138,394 @@ def local_skysub_extract(self, global_sky, sobjs,
...
+            corr_model *= self.caliBrate.flatimages.pixelflat
+            scale_model = self.caliBrate.flatimages.procflat/corr_model
+            scale_model /= blaze_model
+        embed()
+        import astropy.io.fits as fits
+        hdu = fits.PrimaryHDU(scale_model)
+        hdu.writeto('scale_model.fits', overwrite=True)
+        return scale_model
+
+    def build_scaleimg_old(self, ref_slit, trim=10):
```
Comment (profxj) [inline pypeit/reduce.py:1292]: why is it called `old`?

File: pypeit/reduce.py [hunk 16]
```diff
@@ -1081,3 +1138,394 @@ def local_skysub_extract(self, global_sky, sobjs,
...
+            resolution = int(10.0*np.max(self.slits_right-self.slits_left))
+            # Get the regions
+            status, regions = skysub.read_userregions(self.par['reduce']['skysub']['user_regions'],
+                                                      resolution=resolution)
+            # Generate image
+            skymask_init = skysub.generate_mask(self.pypeline, regions, self.slits, self.slits_left, self.slits_right,
+                                                resolution=resolution)
+        return skymask_init
+
+    def run(self, basename=None, ra=None, dec=None, obstime=None,
```
Comment (profxj) [inline pypeit/reduce.py:1449]: Do you have to overload `run`?  I'm guessing yes,
but kindly explain why.   Will be a bit painful to
maintain the two of them..

Comment (rcooke-ast) [reply pypeit/reduce.py]: Agreed... unfortunately, I think this will be needed. Perhaps we will need to be more modular in the original `run()` function, so we can apply certain features but not all. Basically, for IFU, you don't want to find/extract  objects. You just need to extract the entire slit/FOV.

For now, I've removed all of the IFU code from the `reduce.py` code so that it can be the focus of the next PR. This PR is really about bringing all the calibrations up to speed with the rest of the code.

File: pypeit/images/sciencecube.py (hunk 12 above)
Comment (rcooke-ast) [reply pypeit/images/sciencecube.py:23]: I've renamed this, but I'm not sure I put a lot of effort into this code... seems I created it to put in the overlapping pixel/voxel code mainly... Deleting from this PR, and will bring it forward to the next PR.

File: pypeit/images/sciencecube.py (hunk 13 above)
Comment (rcooke-ast) [reply pypeit/images/sciencecube.py:198]: OK - removed (it seems like nothing is calling this code)!

**Comment by rcooke-ast** [comment]
I'll implement some tests shortly, and run the dev suite.

**Comment by rcooke-ast** [comment]
Three tests added - running on the dev suite now

File: pypeit/scripts/skysub_regions.py [hunk 17]
```diff
@@ -52,6 +56,7 @@ def main(args):
 
     # Load the image data
     frame = info.load_frame(sciIdx)
+    print(frame.shape)
```
Comment (kbwestfall) [inline pypeit/scripts/skysub_regions.py:59]: remove?

File: pypeit/spectrographs/keck_kcwi.py [hunk 18]
```diff
@@ -463,3 +512,48 @@ def bpm(self, filename, det, shape=None, msbias=None):
 
         return bpm_img
 
+    def set_wcs(self, hdr):
```
Comment (kbwestfall) [inline pypeit/spectrographs/keck_kcwi.py:515]: Have you played with the `astropy.wcs` package much?  I've only used it to read and construct WCS objects from headers, but I'm wondering if there are some utilities that will make this setup easier, and make sure that it conforms to an expected standard.

File: pypeit/tests/test_skysub.py [hunk 19]
```diff
@@ -0,0 +1,44 @@
+"""
```
Comment (kbwestfall) [inline pypeit/tests/test_skysub.py:1]: Nice!!

File: pypeit/core/arc.py [hunk 20]
```diff
@@ -480,6 +482,10 @@ def get_censpec(slit_cen, slitmask, arcimg, gpm=None, box_rad=3.0, nonlinear_cou
     nspat = arcimg.shape[1]
     spat = np.arange(nspat)
     for islit in range(nslits):
+        # Check if this slit is to be extracted
+        if slitIDs is not None:
+            if islit not in slitIDs: continue
```
[the remark this answers is not in the document]
Comment (rcooke-ast) [reply pypeit/core/arc.py]: Done!

File: pypeit/core/gui/identify.py [hunk 21]
```diff
@@ -243,6 +375,35 @@ def toggle_wavepix(self, toggled=False):
...
 
+    def draw_ghost(self):
+        """Draw tick marks at the location of the ghost
+        """
+        for i in self.gstlines:
+            try:
+                i.remove()
+            except TypeError:
+                i[0].remove()
+        for i in self.gsttexts: i.remove()
```
[the remark this answers is not in the document]
Comment (rcooke-ast) [reply pypeit/core/gui/identify.py]: Done - and there were a couple more in the code like this that I fixed as well.

File: pypeit/core/gui/identify.py [hunk 22]
```diff
@@ -504,21 +740,18 @@ def button_release_callback(self, event):
...
                     # The mouse button was dragged
                     if axisID == 0:
-                        if self._start > self._end:
-                            tmp = self._start
-                            self._start = self._end
-                            self._end = tmp
-                        self.update_regions()
+                        if self._ghostmode:
+                            pass
+                        else:
```
[the remark this answers is not in the document]
Comment (rcooke-ast) [reply pypeit/core/gui/identify.py]: Done!

File: pypeit/core/skysub.py [hunk 23]
```diff
@@ -1116,3 +1108,124 @@ def ech_local_skysub_extract(sciimg, sciivar, fullmask, tilts, waveimg, global_s
...
+    slits_right : ndarray
+        A 2D array containing the pixel coordinates of the right slit edges
+    resolution: int, optional
+        The percentage regions will be scaled to the specified resolution. The
+        resolution should probably correspond to the number of spatial pixels
+        on the slit.
+
+    Returns
+    -------
+        ndarray : Boolean mask containing sky regions
```
[the remark this answers is not in the document]
Comment (rcooke-ast) [reply pypeit/core/skysub.py]: Ok - I think I fixed this one.

File: pypeit/reduce.py [hunk 24]
```diff
@@ -186,7 +182,28 @@ def __init__(self, sciImg, spectrograph, par, caliBrate,
...
+        # Slitmask
+        self.slitmask = self.slits.slit_img(initial=initial, flexure=self.spat_flexure_shift,
+                                           exclude_flag=self.slits.bitmask.exclude_for_reducing)
+        # Now add the slitmask to the mask (i.e. post CR rejection in proc)
+        # NOTE: this uses the par defined by EdgeTraceSet; this will
+        # use the tweaked traces if they exist
+        self.sciImg.update_mask_slitmask(self.slitmask)
+        # For echelle
+        self.spatial_coo = self.slits.spatial_coordinates(initial=initial, flexure=self.spat_flexure_shift)
+        return
```
[the remark this answers is not in the document]
Comment (rcooke-ast) [reply pypeit/reduce.py]: Removed

File: pypeit/reduce.py [hunk 25]
```diff
@@ -1087,3 +1138,385 @@ def local_skysub_extract(self, global_sky, sobjs,
...
+    Child of Reduce for IFU reductions
+
+    See parent doc string for Args and Attributes
+
+    """
+    def __init__(self, sciImg, spectrograph, par, caliBrate, objtype, **kwargs):
+        super(IFUReduce, self).__init__(sciImg, spectrograph, par, caliBrate, objtype, **kwargs)
+        self.initialise_slits(initial=True)
+
+    #
```
[the remark this answers is not in the document]
Comment (rcooke-ast) [reply pypeit/reduce.py]: OK - it'll probably (almost) completely change... but best to keep it in so I don't accidentally delete it!

File: pypeit/scripts/skysub_regions.py (hunk 17 above)
Comment (rcooke-ast) [reply pypeit/scripts/skysub_regions.py]: Good catch - thanks for spotting that 😊

File: pypeit/spectrographs/keck_kcwi.py (hunk 18 above)
Comment (rcooke-ast) [reply pypeit/spectrographs/keck_kcwi.py:515]: I haven't no, but I definitely intend to use it. This is (almost) a straight copy of the KDERP IDL pipeline... I ambitiously made a start on it, but I think I'm a way off that yet!

**Comment by rcooke-ast** [comment]
Unit tests passed. devsuite passes all tests except two:

**NIRES_QL** (can't find MasterSlits)
```
Running pypeit_ql_mos on keck_niresQL  --- FAILED
  File "/Users/rcooke/Software/PypeIt/pypeit/reduce.py", line 195, in initialise_slits
    = self.slits.select_edges(initial=initial, flexure=self.spat_flexure_shift)
AttributeError: 'NoneType' object has no attribute 'select_edges'
```

**GS_HAM_R400_860** (Can't find object)
```
  File "/Users/rcooke/Software/PypeIt/pypeit/pypmsgs.py", line 267, in error
    raise PypeItError(msg)
pypeit.pypmsgs.PypeItError: No matching objects for SPAT1508-SLIT1789-DET01.  Odds are you input the wrong OBJID
```

**Comment by profxj** [comment]
Directory: /data/Projects/Python/PypeIt-development-suite/REDUX_OUT/keck_nires/NIRES/QL
Command line: pypeit_ql_keck_nires /data/Projects/Python/PypeIt-development-suite/RAW_DATA/keck_nires/NIRES s190519_0067.fits s190519_0068.fits
Wrote: /data/Projects/Python/PypeIt-development-suite/REDUX_OUT/keck_nires/NIRES/QL/keck_nires_A/QA/MF_A.html
Running pypeit_ql_mos on keck_niresQL  --- PASSED
[INFO]    :: pypeit.py 763 print_end_time() - Execution time: 1m 2.79s
[INFO]    :: ql_keck_nires.py 99 main() - Data reduction complete
[INFO]    :: ql_keck_nires.py 101 main() - Generating QA HTML

**Comment by rcooke-ast** [comment]
Merging...
